backend/API_standards.md

```diff
@@ -0,0 +1,136 @@
+
+## API Requests
+
+### CRUD, Verbs, and Actions
+A resourceful route provides a mapping between HTTP verbs and URLs to
+controller actions. By convention, each action also maps to a specific CRUD
+operation in a database.
+
+Example for a `Question` resource:
+
+| HTTP Verb | Path             | Service#Function     | Used for                                     |
+| --------- | ---------------- | -------------------- | -------------------------------------------- |
+| GET       | /questions       | questions#index      | list all questions (filterable query string) |
+| POST      | /questions       | questions#create     | create a new question                        |
+| GET       | /questions/:id   | questions#show       | display a specific question                  |
+| PUT       | /questions/:id   | questions#update     | update a specific question                   |
+| DELETE    | /questions/:id   | questions#delete     | delete a specific question                   |
+
+Non CRUD operation will follow the same pattern, example:
+
+| HTTP Verb | Path              | Service#Function     | Used for                                     |
+| --------- | ----------------  | -------------------- | -------------------------------------------- |
+| GET       | /questions/search | questions#search     | search questions (filterable query string)   |
+
+Nested resources will follow the same pattern:
+
+| HTTP Verb | Path                 | Service#Function     | Used for                                     |
+| --------- | -------------------  | -------------------- | -------------------------------------------- |
+| GET       | /company/:id/members | members#index        | list all members for a specific company      |
+
+## API Responses
+
+All endpoints should return a seamless response, independent of the endpoint action or expected result values.
+
+### JSON object response
+Every endpoint should return a JSON object with the appropriate `payload key`, rather than  collection or plain values.
+
+Example:
+
+```json
+{
+  "questions": [{"id": 1, "description": "desc 1"},
+                {"id": 2, "description": "desc 2"}]
+}
+```
+
+### Pagination
+Pagination details would be sent as part of the same JSON object.
+Suggested query string inputs, `page_size` and `page_number`.
+
+Example `GET /questions?page_size=8&page_number=4`
+
+Response Example:
+```json
+{
+  "meta": {
+            "total_pages": 5,
+            "current_page": 1
+          },
+  "questions": [
+                 {"id": 1, "description": "desc 1"},
+                 {"id": 2, "description": "desc 2"}
+               ]
+}
+```
+
+### Error Handling
+All errors should follow the same `key: value` structure, each key represent the specific attribute with error.
+
+Input validation example:
+```json
+{ "email": "Invalid email format",
+  "password": "Invalid length, it should be at least 8 characters"}
```

Comment:
We can add an example when we will sent an array such as parameter and it has errors, wdyt?